File: laboris/reports/list.py
import laboris.config as cfg
import laboris.data as dat
from laboris.reports.util import *

import datetime

def print_entry(task, fmt, ind, spacing):
    data = task
    data['suuid'] = data['uuid'][:8]
    data['due'] = date_abbrev(datetime.datetime.now(), task['dueDate'] if 'dueDate' in task else -1)
    data['title'] = "{:{}}".format(task['title'], spacing[0])
    data['projects'] = "{:{}}".format(" ".join(task['projects']), spacing[1])
    data['tags'] = "{:{}}".format(" ".join(task['tags']), spacing[2])
    data['p'] = ' ' if task['priority'] == -1 else task['priority']
    data['urg'] = data['urg'] if 'urg' in data else 0
    print("{}{}{}\033[0m".format("" if ind % 2 == 1 else cfg.CONFIG['theme']['bg'], cfg.get_urg(data['urg']),
          fmt.format(data)))


def main(args):
    fmt = "{0[id]:2} {0[p]:1} {0[due]:>3} {0[title]} {0[projects]} {0[tags]} {0[urg]:6.3f}"
    if 'all' in args and args['all']:
        fmt = "{0[suuid]:4} {0[title]} {0[projects]} {0[tags]}"
    if 'fmt' in args:
        fmt = args.fmt
    spacing = [5, 8, 4]
    for task in dat.DATA[0]:
        spacing[0] = max(spacing[0], len(task['title']))
        spacing[1] = max(spacing[1], len(" ".join(task['projects'])))
        spacing[2] = max(spacing[2], len(" ".join(task['tags'])))
        if task['times'] and len(task['times'][-1]) == 1:
            print(print_task(task))
        elif 'dueDate' in task and task['dueDate'] and datetime.datetime.fromtimestamp(
                task['dueDate']).date() <= datetime.datetime.now().date():
            print(print_task(task))
    if 'all' in args and args['all']:
        for task in dat.DATA[1]:
            spacing[0] = max(spacing[0], len(task['title']))
            spacing[1] = max(spacing[1], len(" ".join(task['projects'])))
            spacing[2] = max(spacing[2], len(" ".join(task['tags'])))
    data = {
        'id': "ID",
        'uuid': "UUID                            ",
        'suuid': "UUID    ",
        'p': "P",
        'age': "Age",
        'due': "Due",
        'dueShort': "Due",
        'title': "{:{}}".format("Title", spacing[0]),
        'projects': "{:{}}".format("Projects", spacing[1]),
        'tags': "{:{}}".format("Tags", spacing[2]),
        'urg': "Urg",
    }
    tmp_fmt = fmt.replace('{', cfg.CONFIG['theme']['title'] + '{')
    tmp_fmt = tmp_fmt.replace('}', '}\033[0m')
    tmp_fmt = tmp_fmt.replace('f}', '}')
    print(tmp_fmt.format(data))
    for i, task in enumerate(dat.DATA[0]):
        print_entry(task, fmt, i, spacing)
    if 'all' in args and args['all']:
        for i, task in enumerate(dat.DATA[1]):
            print_entry(task, fmt, i, spacing)


def detail(args):
    task = dat.find(args['task'])

File: reports/list.js
const _ = require("lodash");
const moment = require("moment");
const chalk = require("chalk");
const { query, mutation, error } = require("../graphql.js");
const { extract } = require("../util.js");
const { parseFormat, fmtData } = require("../fmt.js");
const { parseDate } = require("../common.js");
const {
  dateDeltaMajor,
  dateDeltaMin,
  dateDeltaFull,
  urgColor
} = require("../print.js");

const fmtKeys = fmtStr => {
  const fmt = _.map(parseFormat(fmtStr), "key");
  var keys = ["urg", "active"];
  if (_.includes(fmt, "id")) {
    keys.push("id");
  }
  if (_.includes(fmt, "title")) {
    keys.push("title");
  }
  if (_.some(["p", "pri", "priority"], o => _.includes(fmt, o))) {
    keys.push("priority");
  }
  if (
    _.some(
      [
        "dueDate",
        "dueDelta",
        "dueDeltaMajor",
        "dueDeltaMin",
        "dueDeltaFull",
        "dueDateTime",
        "dueDateLong",
        "dueDateTimeLong"
      ],
      o => _.includes(fmt, o)
    )
  ) {
    keys.push("dueDate");
  }
  if (
    _.some(
      [
        "entryDate",
        "entryDelta",
        "entryDeltaMajor",
        "entryDeltaMin",
        "entryDeltaFull",
        "entryDateTime",
        "entryDateLong",
        "entryDateTimeLong"
      ],
      o => _.includes(fmt, o)
    )
  ) {
    keys.push("entryDate");
  }
  if (
    _.some(
      [
        "doneDate",
        "doneDelta",
        "doneDeltaMajor",
        "doneDeltaMin",
        "doneDeltaFull",
        "doneDateTime",
        "doneDateLong",
        "doneDateTimeLong"
      ],
      o => _.includes(fmt, o)
    )
  ) {
    keys.push("doneDate");
  }
  if (
    _.some(
      [
        "modifiedDate",
        "modifiedDelta",
        "modifiedDeltaMajor",
        "modifiedDeltaMin",
        "modifiedDeltaFull",
        "modifiedDateTime",
        "modifiedDateLong",
        "modifiedDateTimeLong"
      ],
      o => _.includes(fmt, o)
    )
  ) {
    keys.push("modifiedDate");
  }
  if (_.some(["parents", "parentCount"], o => _.includes(fmt, o))) {
    keys.push("parents{id,title}");
  }
  if (_.some(["children", "childCount"], o => _.includes(fmt, o))) {
    keys.push("children{id,title}");
  }
  if (_.some(["tags", "tagCount"], o => _.includes(fmt, o))) {
    keys.push("tags");
  }
  if (_.some(["hidden"], o => _.includes(fmt, o))) {
    keys.push("hidden");
  }
  return _.join(keys, ",");
};

const genDateKeys = (prefix, value) => {
  if (!value) return {};
  var res = {};
  res[prefix + "Date"] = moment(value).format("YYYY-MM-DD");
  res[prefix + "DateTime"] = moment(value).format("YYYY-MM-DD HH:mm:ss");
  res[prefix + "DateLong"] = moment(value).format("dddd, MMMM Do YYYY");
  res[prefix + "DateTimeLong"] = moment(value).format(
    "dddd, MMMM Do YYYY, h:mm:ss a"
  );
  res[prefix + "Delta"] = dateDeltaMajor(_.now(), value);
  res[prefix + "DeltaMajor"] = dateDeltaMajor(_.now(), value);
  res[prefix + "DeltaMin"] = dateDeltaMin(_.now(), value);
  res[prefix + "DeltaFull"] = dateDeltaFull(_.now(), value);
  return res;
};

const genPrintKeys = task => {
  return {
    title: task.title,
    id: task.id,
    priority: task.priority,
    parents: task.parents
      ? _.join(_.map(task.parents, "title"), " ")
      : undefined,
    parentCount: task.parents ? task.parents.length : undefined,
    children: task.children
      ? _.join(_.map(task.children, "title"), " ")
      : undefined,
    childCount: task.children ? task.children.length : undefined,
    tags: task.tags ? _.join(task.tags, " ") : undefined,
    tagCount: task.tags ? task.tags.length : undefined,
    urg: task.urg,
    hidden: task.hidden,
    active: task.active,
    ...genDateKeys("due", task.dueDate),
    ...genDateKeys("entry", task.entryDate),
    ...genDateKeys("done", task.doneDate),
    ...genDateKeys("modified", task.modifiedDate)
  };
};

module.exports = (args, config) => {
  var queryStr = "filter";
  parsedArgs = {
    priority: extract.priority(args, null),
    parents: extract.parents(args),
    children: extract.children(args),
    tags: extract.tags(args),
    dueBefore:
      extract.dueDate(args) ||
      (args.dueBefore ? parseDate(args.dueBefore) : undefined),
    dueAfter: args.dueAfter ? parseDate(args.dueAfter) : undefined,
    entryBefore: args.entryBefore ? parseDate(args.entryBefore) : undefined,
    entryAfter: args.entryAfter ? parseDate(args.entryAfter) : undefined,
    modifiedBefore: args.modifiedBefore
      ? parseDate(args.modifiedBefore)
      : undefined,
    modifiedAfter: args.modifiedAfter
      ? parseDate(args.modifiedAfter)
      : undefined,
    hidden: args.hidden || undefined
  };
  if (
    parsedArgs.priority !== null ||
    parsedArgs.parents.length !== 0 ||
    parsedArgs.children.length !== 0 ||
    parsedArgs.tags.length !== 0 ||
    parsedArgs.dueBefore !== undefined ||
    parsedArgs.dueAfter !== undefined ||
    parsedArgs.entryBefore !== undefined ||
    parsedArgs.entryAfter !== undefined ||
    parsedArgs.modifiedBefore !== undefined ||
    parsedArgs.modifiedAfter !== undefined ||
    parsedArgs.hidden !== undefined
  ) {
    queryStr += "(";
    for (const key in parsedArgs) {
      if (typeof parsedArgs[key] === "object") {
        if (parsedArgs[key] !== null && parsedArgs[key].length !== 0)
          queryStr += key + ":" + JSON.stringify(parsedArgs[key]) + ",";
      } else if (parsedArgs[key] !== null && parsedArgs[key] !== undefined) {
        queryStr += key + ":" + parsedArgs[key] + ",";
      }
    }
    queryStr = queryStr.slice(0, -1);
    queryStr += ")";
  }
  queryStr += "{" + fmtKeys(config.get("listFormat")) + "}";
  fmt = parseFormat(config.get("listFormat"));
  query(queryStr, config)
    .then(data => {
      data.filter = _.map(
        data.filter.sort((a, b) => b.urg - a.urg),
        genPrintKeys
      );
      printData = {};
      fmt.forEach(fmt => {
        printData[fmt.key] = [];
        data.filter.forEach(task => {
          printData[fmt.key].push(fmtData(fmt, task));
        });
        const longest = _.max(printData[fmt.key], str => str.length).length;
        if (longest === 0) delete printData[fmt.key];
        else
          printData[fmt.key] = _.map(printData[fmt.key], str => {
            if (fmt.align === "<") return str.padEnd(longest);
            else return str.padStart(longest);
          });
      });
      for (const i in data.filter) {
        var line = "";
        for (const key in printData) {
          line += "  " + printData[key][i];
        }
        if (i % 2 === 0)
          console.log(
            urgColor(data.filter[i].urg, data.filter[i].active)(line)
          );
        else
          console.log(
            chalk.bgBlack(
              urgColor(data.filter[i].urg, data.filter[i].active)(line)
            )
          );
      }
    })
    .catch(err => error);
};
